### PR TITLE
console: add receive syscalls

### PR DESCRIPTION
**Pull request by daniel-scs**
### Pull Request Overview

- libtock now includes getnstr(), getnstr_async(), getch()

### Testing Strategy

- An example app, "console", demonstrates that getch() works
- The boards imix and hail run the "console" app successfully: you can run `tockloader listen` and it will echo back characters you type

### TODO or Help Wanted

- Further testing is required to exercise the other syscalls
- Further thought is required on how to multiplex app requests

### Documentation Updated

- [ ] Kernel: Updated the relevant files in `/docs`, or no updates are required.
- [X] Userland: Added/updated the application README, if needed.

### Formatting

- [ ] Ran `make formatall`.

### CODE REVIEW

**Comment by daniel-scs** [comment]
This implementation is not quite complete -- I cut some corners to get the basics working -- so perhaps we should call this Blocked for now and save the reviews until I've polished it off.

**Comment by bradjc** [comment]
Do need to update https://github.com/helena-project/tock/blob/master/doc/syscalls/00001_console.md at some point.

File: capsules/src/console.rs [hunk 1]
```diff
@@ -162,6 +175,39 @@ impl<'a, U: UART> Console<'a, U> {
...
+                    // For simplicity, impose a small maximum receive length
+                    // instead of doing incremental reads
+                    ReturnCode::EINVAL
+                }
+                else {
+                    // Note: We have ensured above that rx_buffer is present
+                    app.read_len = read_len;
+                    self.rx_buffer.take().map(|buffer| {
+                        self.rx_in_progress.set(Some(app_id));
+                        self.uart.receive(buffer, app.read_len);
```
Comment (niklasad1) [inline capsules/src/console.rs]: why not `self.uart,receive(&buffer[0..app.read_len])`? Passing the length explicitly seems unnecessary to me!

Comment (alevy) [reply capsules/src/console.rs]: to receive _into_ that buffer, it has to be mutable. for it to be mutable, it _has_ to be returned to the caller through the callback. if we pass in a sliced buffer, there will be no way to recover the complete slice in the callback.

So far, the solution we've been using is to pass around the entire buffer along with an explicit length.

We could also have a special buffer type that keeps track of both a current slice but also has the full origin and length so the original buffer can be recovered. But that's probably best not to deal with in this PR.

Comment (niklasad1) [reply capsules/src/console.rs]: Yeah, right never mind didn't think about that :frowning_woman:

**Pull request merged**